### ISSUE TITLE
Trouble shooting through grates

### ISSUE THREAD

**Issue by CHILLYBUS** [dxx-rebirth/dxx-rebirth#588]
#### Operating System Environment
* Microsoft Windows
  * [ ] Microsoft Windows XP
  * [ ] Microsoft Windows Vista
  * [ ] Microsoft Windows 7
  * [ ] Microsoft Windows 8
  * [ ] Microsoft Windows 8.1
  * [X] Microsoft Windows 10

#### CPU environment
* [ ] x86 (32-bit Intel/AMD)
* [X] x86\_64 (64-bit Intel/AMD)
* [ ] ARM (32-bit)
* [ ] ARM64 (64-bit; sometimes called AArch64)

#### Game environment
Any location with transparent grate assets.

### Description
Grates typically can be shot through in older builds of Descent II - the original hitboxes aren't particularly exacting, but are approximate enough that any weapons fire can be reasonably sent through them provided the player's aim is good enough. That isn't the case in recent versions of D2X-Rebirth. I've done some cursory tests myself with recent builds after seeing some players have trouble during some Twitch streams of the game, and encounter roughly the same issues.

Some streamers report an identical issue in D1X, and other players report this issue not occurring for them at all.

#### Regression status
This issue was not present in early builds of Rebirth 60/61, but is present in late-2020 and early-2021 builds of the game, if that's helpful. 

v0.61 0.60.0-beta2-709-GF0A47C358BA6* and earlier appear to be working. Still unsure which exact build this issue starts with.

### Steps to Reproduce
Fly to a grate, try shooting through the grate. Best demo of the problem is shooting through wide grates - the vulcan window into the red key room on level 1 of D2: Counterstrike is the best example, but switches behind tight grates in that level are good tests of the issue too since those need to be shoot-through-able too.

* Engines affected:
  * [ ] D1X-Rebirth
  * [X] D2X-Rebirth

* Game modes affected:
  * [X] Single player
  * [ ] Multiplayer competitive
    * [ ] Anarchy
    * [ ] Team Anarchy
    * [ ] Robo-Anarchy
    * [ ] Capture the Flag
    * [ ] Bounty
  * [?] Multiplayer cooperative

Haven't tested cooperative play to see if this is present.

**Comment by vLKp** [comment]
> Some streamers report an identical issue in D1X, and other players report this issue not occurring for them at all.

This ticket is the first I have heard of this problem.  Prior to now, no one reported it in any venue I monitor.  I think most of the collision processing code is shared between D1/D2 at this point.

> This issue was not present in early builds of Rebirth 60/61, but is present in late-2020 and early-2021 builds of the game, if that's helpful.

It's better than some reports I have received, but not precise enough that it is likely to be useful, unfortunately.  I like getting date-based regression reports, but to be helpful, I need either:

- specific commits
- links to specific builds, from which I can determine the commits
- specific days identified as good/bad

> Best demo of the problem is shooting through wide grates - the vulcan window into the red key room on level 1 of D2: Counterstrike is the best example ...

This, on the other hand, is likely to be very helpful.  Reproductions in stock levels are nice because everybody has them, and a reproduction on level 1 is even better, since it's easier to cold start a campaign and get to the reproduction site.

Could you describe this reproduction in a bit more detail?  The red key is located in the Weapons Testing Area, at the base of a lava fall.  I am uncertain what you mean by the "vulcan window" though.  Do you mean to fly to where the player obtains a vulcan cannon, open the secret door that offers access to the ammo rack, and then try to fire through the grating that separates the ammo rack from the red key room?  While floating in the red key room, I was able to use a Vulcan cannon to kill the PIG that sits in the room with the ammo rack, and did not notice any issues.  I then flew out and around to the ammo rack room.  I stopped short of picking up the rack, floated toward the floor, and successfully killed a PEST using a Vulcan cannon.  In all cases, I used a developer feature (DELETE+P in builds where it exists) to pause robot processing, so my targets stayed exactly where the original level placed them, and did not return fire.

So far, I do not experience the reported problem.  Are there any recorded streams or published demos that exhibit the issue?  The data in those will not be sufficient to debug it, but may show me a way in which my failed test today differed from the successful reproductions other users have seen.

> Haven't tested cooperative play to see if this is present.

That is fine.  If the issue reproduces in single player, that is both sufficient to merit fixing it and easier to work on than in multiplayer.  Other people might find it interesting to know whether cooperative is affected, but I do not need that answered in order to address this ticket.

**Comment by KynikossDragonn** [comment]
For what it's worth I also have this strange collision detection anomaly happen where player vulcan shots can't seem to squeeze through grate textures in Descent 2, and lasers have a even harder time, exploding even though I'm clearly seeing them go through the tiny hole(s) in the texture.

I'm running 0.60.0-beta2-1313-g809af2987594 as of writing this response.

**Comment by AlumiuN** [comment]
I can confirm that I have the issue as of 809af29; the aforementioned window between the red key room and the secret that has the ammo rack in D2 level 1 seems to exhibit this reliably for me (being completely solid from either room). I recorded a demo with this behaviour occuring.

[grate_test.zip](https://github.com/dxx-rebirth/dxx-rebirth/files/6368769/grate_test.zip)

I'm going to do some testing of my own to try and narrow down the issue this afternoon. I'm still trying to figure out where the function that actually tests a projectile collision against a transparent wall is located (I had tracked it down previously to investigate a bug in a user mission that turned out to be a presumably intended reimplemented bug from D1, long story, but I can't remember where it was now), but the problem isn't necessarily there either.

**Comment by 4C1T** [comment]
I have also been experiencing this bug and have done a bunch of tests trying to figure out what's going on. Here's what I've discovered so far.

As anyone who's played around with the level editor will know, walls have a Texture 1 and an optional Texture 2, and textures have two transparent colors: Transparent and See-Thru. This bug seems to affect, collectively as a group, walls with all four of the following properties:

1. The wall can't be flown through. (This excludes illusory walls and fully open doors.)
2. The wall doesn't have a Texture 2. (This excludes most doors and some grates.)
3. The wall's Texture 1 hasn't been edited. (I used DLE to edit textures.)
4. The wall's Texture 1 contains at least some pixels colored Transparent. (Pixels colored See-Thru don't matter.)

Examples of walls of this kind include force fields, Guide Bot and prison doors, many types of grates, and particular types of doors that are on particular frames of their opening/closing animation — basically, most walls that you can shoot through but can't fly through. This bug also seems to affect cloaked walls, although I'm less sure what properties are required for this to be the case.

So here's what seems to be happening:

Walls of the kind described above have the permeability of, not their own texture, but the texture of whichever wall of this kind happened to have been fired upon first. This affects all projectiles, not just those fired by the player.

For example, if the first thing I do after loading Ahayweh Gate is fly through the right-hand door and shoot at the IT Droids through the grates (metl146), then for the rest of that level, other grates with different textures will behave as though they have the texture metl146 when it comes to their permeability to weapons fire. However, if the Guide Bot happened to have fired a flare into its cell door (door53) as I was passing by on my way to those grates, then by the time I reach them, those same grates and others will have the permeability of door53, not metl146.

To be clear, only the _texture_ of the initially-fired-upon wall is used in this bugged permeability mapping, not its texture _alignment._ The alignment is that of the wall actively being fired upon, which is exactly as it should be. Also, I should reiterate that all of this only applies to walls with the properties that I described above.

This bug is 100% consistent and affects both D1X-Rebirth and D2X-Rebirth, although it's harder to notice in D1X-Rebirth because fewer walls have the required properties. The bug does _not_ persist across levels, nor across saving and reloading during a level, however nothing I do _inside_ a level seems to reset the bugged permeability once it has locked in on a particular texture. It also seems to be completely unaffected by changes to any graphics settings.

**Comment by AlumiuN** [comment]
A quick test confirms this behaviour on my end too, provided I don't shoot the secret door in the roof of the tunnel to the left of the starting area (as doing that results in all grates being completely solid); opening the secret door to the ammo rack doesn't cause the same issue.

**Comment by AlumiuN** [comment]
I've found the issue, although I'm not exactly sure of its cause. The issue was introduced in commit 61f186b, specifically the change to check_trans_wall in similar/main/fvi.cpp. The reason I'm not sure of its cause is that the code appears to be correct, and extricating the code in the closure (lines 1209 through 1214) out to a function makes it work correctly and fixes the issue. My theory is that the closure's result is being optimised by the compiler to be the same for every call after the first until certain points (loading a saved game, changing level, returning to the main menu), which even with my limited understanding of closures seems incorrect.

**Comment by 4C1T** [comment]
@AlumiuN That's because your shot contains multiple projectiles and one is hitting the secret door ever so slightly after the first. The first projectile triggers the door to open and the second hits it on Frame 1 of its opening animation. Doors are a bit weird with this bug. Even though Frame 1 doesn't actually contain any Transparent pixels, it still seems to count for some reason as satisfying property 4 in my list. This means that, at the moment the second projectile hits the door, all four properties are satisfied and so permeability gets locked in on the completely opaque Frame 1 texture. To prevent this from happening, try using flares to open secret doors and avoid shooting them while they're in the process of opening or closing.

**Comment by AlumiuN** [comment]
> @AlumiuN That's because your shot contains multiple projectiles and one is hitting the secret door ever so slightly after the first. The first projectile triggers the door to open and the second hits it on Frame 1 of its opening animation. Doors are a bit weird with this bug. Even though Frame 1 doesn't actually contain any Transparent pixels, it still seems to count for some reason as satisfying property 4 in my list. This means that, at the moment the second projectile hits the door, all four properties are satisfied and so permeability gets locked in on the completely opaque Frame 1 texture. To prevent this from happening, try using flares to open secret doors and avoid shooting them while they're in the process of opening or closing.

Yeah, I realised that a few minutes ago while I was typing up the previous post; I suspect the reason the secret door to the ammo rack didn't trigger it is because I'd picked up and autoswitched to the Vulcan at that point.

**Comment by vLKp** [comment]
Thank you all for the research and analysis, and @AlumiuN in particular for finding the first bad commit and a proposed fix.  With that fix available, I can explain why it works.  The closure is the problem, but not for the reason that @AlumiuN stated.  Rather, the issue is that:

- I allowed the closure to deduce its return type, and the return type is deduced as `grs_bitmap`, _not_ `const grs_bitmap &`.  On its own, this just makes the code slightly less efficient than it ought to be.
- The return value of the closure is bound to `const grs_bitmap &rbm`.  A special rule in C++ permits binding a `const T &` to an anonymous instance of `T`, so this creates a _copy_ of the chosen bitmap on the stack, and points `rbm` to it.  This allowed the above mistake to be valid code, and to have well-defined runtime behavior.  (There was no dangling reference to a destroyed object.)
- `rle_expand_texture` maintains a cache of recently expanded textures, and uses as its cache key the *address* of the input `grs_bitmap` that it is expanding.  For persistent textures, as stored in `GameBitmaps`, this is fine.  For a transient texture stored on the stack of `check_trans_wall`, this is bad, because future calls of `check_trans_wall` will likely have the same call stack, and stack layout, resulting in every call of `check_trans_wall` using the same memory address for the result of the closure.  Since `rle_expand_texture` compares memory addresses, rather than contents, this makes every bitmap checked by `check_trans_wall` "the same" as far as `rle_expand_texture` is concerned, so the result of the previous call is reused, rather than repeating the decoding of the bitmap.  To the player, each wall looks as it should.  To `check_trans_wall`, each wall looks like the first wall you checked for transparency, which is why shooting a non-permeable wall first causes later permeable walls to act impermeable.  (This is also why my attempt to reproduce originally failed.  I happened not to shoot any non-permeable walls before shooting the test wall, so `rle_expand_texture` was called for the first time with the permeable wall, and decoded it properly.)

When the RLE cache is flushed, as can happen on starting a new level, the bad entry is removed.  This returns the game to its starting state, allowing the next use of `check_trans_wall` to RLE decode the tested wall correctly.  If the first post-flush call of `check_trans_wall` is to a non-permeable wall, this will just re-cache another bad entry.  If the first post-flush call is for a permeable wall, then the desired result can occur.

Since this is a problem with incorrectly reusing cached RLE decode results, a wall texture which is not RLE encoded will work correctly even with the bug in place, because such a wall will not call `rle_expand_texture` to generate a temporary texture against which the hit test is performed.

The change proposed by @AlumiuN works not because it eliminates the closure, but because it forces the correct return type, so `rbm` is bound to the persistent copy of the texture in `GameBitmaps`, not to a transient copy on the stack of `check_trans_wall`.  Knowing this, it is possible to fix the code more simply, by modifying the definition of the closure to force a return type of `const grs_bitmap &` instead of allowing the compiler to deduce `grs_bitmap`.

**Comment by AlumiuN** [comment]
Oh wow, I was not going to figure that one out myself. Nicely done :)

Although it's only vaguely relevant, especially now that the issue is fixed - does that special rule for the const reference mean that that temporary will be kept alive until the reference drops out of scope too, or will the compiler error upon an attempt to return a temporary bound in such a fashion out of its scope?

**Comment by KynikossDragonn** [comment]
I can confirm the fix works on my end, especially since I use Clang and not GCC to build things with when I'm given the option to.

**Comment by vLKp** [comment]
> does that special rule for the const reference mean that that temporary will be kept alive until the reference drops out of scope too

The temporary persists until the end of scope of the original `const &` bound to it, with some quirks.  Creating additional const references of longer duration does not further extend the lifetime, so those references will become dangling when the initial reference goes out of scope.  See https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary for the full details.

Yes, this feature, though useful, can definitely lead to bugs with dangling references.

> especially since I use Clang and not GCC to build things with when I'm given the option to.

Thank you for the confirmation.  To be clear, the closure as written was wrong given how RLE caching works, and conforming compilers are within their rights to generate a program that exhibits the overzealous caching.  It is possible, but unlikely, that a compiler might have generated a program where the memory location of the temporary varies across calls to `check_trans_wall`, with the result that the RLE caching might not incorrectly reuse bitmaps.  However, the obvious way of generating the code is to do as happened here.  I considered changing the RLE caching to use the memory address of the encoded texture's data, rather than the memory address of the `grs_bitmap` that owns the encoded data, but decided not to make that change at this time.  It should be safe, but since I never intended for the bitmap to be copied to a stack local, fixing that was a simpler and less invasive choice.

Based on the confirmation above, and the fact that the failure is now well understood, I am closing this issue.  If someone can reproduce this with the fixed code, I am happy to reopen and dig deeper.

**Comment by Estuways** [comment]
Hello folks,
I'm not sure if this belongs here, but I should have the most recent d2x version (downloaded April 2022) and I'm experiencing the game-breaking bug that I can't shoot through force fields. On level 2, this makes completing the level impossible since you have to shoot through the force field behind which the red key is in order to get it. It's impossible with any weapon. I'm using Windows 10. Does anyone have an idea how to fix this?

**Comment by AlumiuN** [comment]
I can't replicate this on the most recent build from the website (March 22); there was a build released this time last year (roughly) that did exhibit this problem, are you sure you didn't download that build instead accidentally?

**Comment by vLKp** [comment]
@Estuways wrote:

> I'm not sure if this belongs here

While I do invite people to report regressions and incomplete fixes in the original ticket, your report is unfortunately insufficient to reexamine the problem.  You wrote when you downloaded a bad build, but did not provide a link to it, nor did you provide its version string.  Thus, as @AlumiuN hints, it is possible you downloaded an old and bad build, and we cannot confirm or reject that from the information shown.  Please provide either an exact download URL so that I can fetch the same build you did, or provide the game's version string (0.60.0-beta2-NNNN-gXXXX for varying N and X) so that I can confirm the build you downloaded is expected to work.

**Comment by Estuways** [comment]
> your report is unfortunately insufficient

Thank you both for your fast replies. I'm not very familiar with version strings and all that, apologies.
I downloaded "dx2rebirth20210224.zip" from https://www.gamepressure.com/download.asp?ID=77583. I suppose they haven't updated the game version for a while.

I've now downloaded the most recent version from the proper page and it works as it should. I should have done that do begin with. Thank you!
